<a href="https://colab.research.google.com/github/pag6521/alist-doc/blob/main/Google_Colab_vs_GitHub_ubuntu_runner.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a href="https://github.com/groda/big_data/blob/master/README.md"><div><img src="https://github.com/groda/big_data/blob/master/logo_bdb.png?raw=true" align=right width="90" alt="Logo Big Data for Beginners"></div></a>
# Google Colab vs. GitHub Ubuntu Runner


        

I've noticed that the Big Data notebooks in the BDb repo run quite fast when [being tested by GitHub actions](https://github.com/groda/big_data/actions/workflows/run-notebooks.yml) compared to their runtimes on Google Colaboratory. You can find some timings in: [action_log.txt](https://github.com/groda/big_data/blob/367429dca1faeefc7c2d2dbfda023609a683c055/action_log.txt).

So I compared the standard Google Colab virtual machine with GitHub's "ubuntu:latest" (**\***) runner and found out that not only the GitHub runner has $15$GB RAM ($3$GB more than Google Colab) but it also has double the amount of virtual CPU's ($4$ versus the $2$ of Google Colab).

(**\***) As of December 2024, GitHub's "ubuntu-latest" image [is soon going to be replaced by "ubuntu-24.04"](https://github.com/actions/runner-images/issues/10636) (currently it's 22.04) so I'll need to substitute "ubuntu-22.04" to "ubuntu-latest" in GitHub since Colab is currently using "ubuntu-22.04".

Below is a comparison table.

<table style="border: 1px solid black;vertical-align: top;">
<tr>
<th>
<th>Google Colab
<th>GitHub ubuntu runner
<tr>
<td><pre>free -h</pre>
<td>
<pre>               total        used        free      shared  buff/cache   available
Mem:            12Gi       827Mi       7.1Gi       1.0Mi       4.7Gi        11Gi
Swap:             0B          0B          0B
</pre>
 <td>
<pre>
               total        used        free      shared  buff/cache   available
Mem:            15Gi       683Mi        13Gi        23Mi       1.0Gi        14Gi
Swap:          4.0Gi          0B       4.0Gi
</pre>
<tr>
<tr>
<td>
<pre>lscpu</pre>
<td>
<pre>
Architecture:             x86_64
  CPU op-mode(s):         32-bit, 64-bit
  Address sizes:          46 bits physical, 48 bits virtual
  Byte Order:             Little Endian
CPU(s):                   2
  On-line CPU(s) list:    0,1
Vendor ID:                GenuineIntel
  Model name:             Intel(R) Xeon(R) CPU @ 2.20GHz
    CPU family:           6
    Model:                79
    Thread(s) per core:   2
    Core(s) per socket:   1
    Socket(s):            1
    Stepping:             0
    BogoMIPS:             4399.99
    Flags:                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 cl
                          flush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc re
                          p_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3
                           fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand
                           hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp
                          fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx sm
                          ap xsaveopt arat md_clear arch_capabilities
Virtualization features:  
  Hypervisor vendor:      KVM
  Virtualization type:    full
Caches (sum of all):      
  L1d:                    32 KiB (1 instance)
  L1i:                    32 KiB (1 instance)
  L2:                     256 KiB (1 instance)
  L3:                     55 MiB (1 instance)
NUMA:                     
  NUMA node(s):           1
  NUMA node0 CPU(s):      0,1
Vulnerabilities:          
  Gather data sampling:   Not affected
  Itlb multihit:          Not affected
  L1tf:                   Mitigation; PTE Inversion
  Mds:                    Vulnerable; SMT Host state unknown
  Meltdown:               Vulnerable
  Mmio stale data:        Vulnerable
  Reg file data sampling: Not affected
  Retbleed:               Vulnerable
  Spec rstack overflow:   Not affected
  Spec store bypass:      Vulnerable
  Spectre v1:             Vulnerable: __user pointer sanitization and usercopy barriers only; no swa
                          pgs barriers
  Spectre v2:             Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BH
                          I: Vulnerable (Syscall hardening enabled)
  Srbds:                  Not affected
  Tsx async abort:        Vulnerable
  </pre>
<td>
<pre>
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             4
On-line CPU(s) list:                0-3
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 7763 64-Core Processor
CPU family:                         25
Model:                              1
Thread(s) per core:                 2
Core(s) per socket:                 2
Socket(s):                          1
Stepping:                           1
BogoMIPS:                           4890.85
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep
                                    mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext
                                    fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc
                                    cpuid extd_apicid aperfmperf pni pclmulqdq
                                    ssse3 fma cx16 pcid sse4_1 sse4_2
                                    movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm
                                    cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext invpcid_single
                                    vmmcall fsgsbase bmi1 avx2 smep bmi2 erms
                                    invpcid rdseed adx smap clflushopt
                                    clwb sha_ni xsaveopt xsavec xgetbv1 xsaves clzero xsaveerptr rdpru arat npt
                                    nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter
                                    pfthreshold v_vmsave_vmload umip vaes vpclmulqdq rdpid fsrm
Virtualization:                     AMD-V
Hypervisor vendor:                  Microsoft
Virtualization type:                full
L1d cache:                          64 KiB (2 instances)
L1i cache:                          64 KiB (2 instances)
L2 cache:                           1 MiB (2 instances)
L3 cache:                           32 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-3
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Vulnerable: Safe RET, no microcode
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines; STIBP disabled;
                                    RSB filling; PBRSB-eIBRS Not affected;
                                    BHI Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected
</pre>
</table>

## Data collection on Colab

Get information on the processing units (CPUs) with [`lscpu`](https://unix.stackexchange.com/questions/468766/understanding-output-of-lscpu):

In [1]:
!lscpu

Architecture:             x86_64
  CPU op-mode(s):         32-bit, 64-bit
  Address sizes:          46 bits physical, 48 bits virtual
  Byte Order:             Little Endian
CPU(s):                   2
  On-line CPU(s) list:    0,1
Vendor ID:                GenuineIntel
  Model name:             Intel(R) Xeon(R) CPU @ 2.00GHz
    CPU family:           6
    Model:                85
    Thread(s) per core:   2
    Core(s) per socket:   1
    Socket(s):            1
    Stepping:             3
    BogoMIPS:             4000.45
    Flags:                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge m
                          ca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht sysc
                          all nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xt
                          opology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq
                           ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt
                           aes xsave avx f16c rdrand hypervisor 

Show the amount of total/free/used memory in human-readable format (option `-h`):

In [2]:
!free -h

               total        used        free      shared  buff/cache   available
Mem:            12Gi       777Mi       8.9Gi       1.0Mi       3.0Gi        11Gi
Swap:             0B          0B          0B


In [3]:
!cat /etc/os-release

PRETTY_NAME="Ubuntu 22.04.4 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.4 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy


## Data collection on GitHub

In order to collect information on the GitHub runner I created an [_action_](https://docs.github.com/en/actions/writing-workflows/quickstart) defined in the YAML file [get_info_on_runner.yml](https://github.com/groda/big_data/blob/master/.github/workflows/get_info_on_runner.yml):

```bash
name: Get info on runner
run-name: Runner info
on:
  # needs to be started manually (by clicking on it)
  workflow_dispatch:
jobs:
  Runner-Info-RAM-CPU:
    runs-on: ubuntu-latest
    steps:
      - run: lscpu
      - run: free -h
      - run: cat /etc/os-release
      - run: echo "🍏 This job's status is ${{ job.status }}."
```

**Status:** Dec. 02, 2024

(these things change over time)

# Task
分析並優化歸納整理檔案 "/content/nba_logreg1.csv" 的資料。

## 載入資料

### Subtask:
使用 pandas 讀取 `/content/nba_logreg1.csv` 檔案到 DataFrame 中。


**Reasoning**:
Import the pandas library and read the CSV file into a DataFrame.



In [4]:
import pandas as pd

df = pd.read_csv('/content/nba_logreg1.csv')

**Reasoning**:
Display the first few rows of the DataFrame to verify the data has been loaded correctly.



In [5]:
display(df.head())

Name  GP   MIN  PTS  FGM  FGA   FG%  3P Made  3PA   3P%  ...  \
0   Brandon Ingram  36  27.4  7.4  2.6  7.6  34.7      0.5  2.1  25.0  ...   
1  Andrew Harrison  35  26.9  7.2  2.0  6.7  29.6      0.7  2.8  23.5  ...   
2   JaKarr Sampson  74  15.3  5.2  2.0  4.7  42.2      0.4  1.7  24.4  ...   
3      Malik Sealy  58  11.6  5.7  2.3  5.5  42.6      0.1  0.5  22.6  ...   
4      Matt Geiger  48  11.5  4.5  1.6  3.0  52.4      0.0  0.1   0.0  ...   

   FTA   FT%  OREB  DREB  REB  AST  STL  BLK  TOV  TARGET_5Yrs  
0  2.3  69.9   0.7   3.4  4.1  1.9  0.4  0.4  1.3            0  
1  3.4  76.5   0.5   2.0  2.4  3.7  1.1  0.5  1.6            0  
2  1.3  67.0   0.5   1.7  2.2  1.0  0.5  0.3  1.0            0  
3  1.3  68.9   1.0   0.9  1.9  0.8  0.6  0.1  1.0            1  
4  1.9  67.4   1.0   1.5  2.5  0.3  0.3  0.4  0.8            1  

[5 rows x 21 columns]

## 資料概覽

### Subtask:
顯示 DataFrame 的前幾行、列名、資料類型以及缺失值資訊，以初步了解資料結構和內容。


**Reasoning**:
Display the head, column names, data types, and non-null counts of the DataFrame to understand its structure and content.



In [6]:
display(df.head())
display(df.columns)
display(df.info())

Name  GP   MIN  PTS  FGM  FGA   FG%  3P Made  3PA   3P%  ...  \
0   Brandon Ingram  36  27.4  7.4  2.6  7.6  34.7      0.5  2.1  25.0  ...   
1  Andrew Harrison  35  26.9  7.2  2.0  6.7  29.6      0.7  2.8  23.5  ...   
2   JaKarr Sampson  74  15.3  5.2  2.0  4.7  42.2      0.4  1.7  24.4  ...   
3      Malik Sealy  58  11.6  5.7  2.3  5.5  42.6      0.1  0.5  22.6  ...   
4      Matt Geiger  48  11.5  4.5  1.6  3.0  52.4      0.0  0.1   0.0  ...   

   FTA   FT%  OREB  DREB  REB  AST  STL  BLK  TOV  TARGET_5Yrs  
0  2.3  69.9   0.7   3.4  4.1  1.9  0.4  0.4  1.3            0  
1  3.4  76.5   0.5   2.0  2.4  3.7  1.1  0.5  1.6            0  
2  1.3  67.0   0.5   1.7  2.2  1.0  0.5  0.3  1.0            0  
3  1.3  68.9   1.0   0.9  1.9  0.8  0.6  0.1  1.0            1  
4  1.9  67.4   1.0   1.5  2.5  0.3  0.3  0.4  0.8            1  

[5 rows x 21 columns]

Index(['Name', 'GP', 'MIN', 'PTS', 'FGM', 'FGA', 'FG%', '3P Made', '3PA',
       '3P%', 'FTM', 'FTA', 'FT%', 'OREB', 'DREB', 'REB', 'AST', 'STL', 'BLK',
       'TOV', 'TARGET_5Yrs'],
      dtype='object')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1340 entries, 0 to 1339
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Name         1340 non-null   object 
 1   GP           1340 non-null   int64  
 2   MIN          1340 non-null   float64
 3   PTS          1340 non-null   float64
 4   FGM          1340 non-null   float64
 5   FGA          1340 non-null   float64
 6   FG%          1340 non-null   float64
 7   3P Made      1340 non-null   float64
 8   3PA          1340 non-null   float64
 9   3P%          1329 non-null   float64
 10  FTM          1340 non-null   float64
 11  FTA          1340 non-null   float64
 12  FT%          1340 non-null   float64
 13  OREB         1340 non-null   float64
 14  DREB         1340 non-null   float64
 15  REB          1340 non-null   float64
 16  AST          1340 non-null   float64
 17  STL          1340 non-null   float64
 18  BLK          1340 non-null   float64
 19  TOV   

None

## 描述性統計

### Subtask:
計算數值型欄位的描述性統計量 (如均值、標準差、最小值、最大值等)，並顯示類別型欄位的計數。


**Reasoning**:
Calculate and display descriptive statistics for numerical columns and value counts for categorical columns as per the instructions.



In [7]:
display(df.describe())
display(df['Name'].value_counts())
display(df['TARGET_5Yrs'].value_counts())

GP          MIN          PTS          FGM          FGA  \
count  1340.000000  1340.000000  1340.000000  1340.000000  1340.000000   
mean     60.414179    17.624627     6.801493     2.629104     5.885299   
std      17.433992     8.307964     4.357545     1.683555     3.593488   
min      11.000000     3.100000     0.700000     0.300000     0.800000   
25%      47.000000    10.875000     3.700000     1.400000     3.300000   
50%      63.000000    16.100000     5.550000     2.100000     4.800000   
75%      77.000000    22.900000     8.800000     3.400000     7.500000   
max      82.000000    40.900000    28.200000    10.200000    19.800000   

               FG%      3P Made          3PA          3P%          FTM  \
count  1340.000000  1340.000000  1340.000000  1329.000000  1340.000000   
mean     44.169403     0.247612     0.779179    19.308126     1.297687   
std       6.137679     0.383688     1.061847    16.022916     0.987246   
min      23.800000     0.000000     0.000000     0.000000     0.000000   
25%      40.200000     0.000000     0.000000     0.000000     0.600000   
50%      44.100000     0.100000     0.300000    22.400000     1.000000   
75%      47.900000     0.400000     1.200000    32.500000     1.600000   
max      73.700000     2.300000     6.500000   100.000000     7.700000   

               FTA          FT%         OREB         DREB          REB  \
count  1340.000000  1340.000000  1340.000000  1340.000000  1340.000000   
mean      1.821940    70.300299     1.009403     2.025746     3.034478   
std       1.322984    10.578479     0.777119     1.360008     2.057774   
min       0.000000     0.000000     0.000000     0.200000     0.300000   
25%       0.900000    64.700000     0.400000     1.000000     1.500000   
50%       1.500000    71.250000     0.800000     1.700000     2.500000   
75%       2.300000    77.600000     1.400000     2.600000     4.000000   
max      10.200000   100.000000     5.300000     9.600000    13.900000   

               AST          STL          BLK          TOV  TARGET_5Yrs  
count  1340.000000  1340.000000  1340.000000  1340.000000  1340.000000  
mean      1.550522     0.618507     0.368582     1.193582     0.620149  
std       1.471169     0.409759     0.429049     0.722541     0.485531  
min       0.000000     0.000000     0.000000     0.100000     0.000000  
25%       0.600000     0.300000     0.100000     0.700000     0.000000  
50%       1.100000     0.500000     0.200000     1.000000     1.000000  
75%       2.000000     0.800000     0.500000     1.500000     1.000000  
max      10.600000     2.500000     3.900000     4.400000     1.000000

Name
Charles Smith       9
Charles Jones       6
Dee Brown           4
Michael Smith       4
Mark Davis          4
                   ..
Archie Goodwin      1
John Bagley         1
Russ Schoene        1
Richard Anderson    1
Dave Batton         1
Name: count, Length: 1294, dtype: int64

TARGET_5Yrs
1    831
0    509
Name: count, dtype: int64

## 資料清理

### Subtask:
根據概覽和統計結果，處理缺失值、異常值或重複項。


**Reasoning**:
Check for and remove duplicate rows, then handle missing values in the '3P%' column by dropping rows with missing values, and finally display the number of remaining missing values to confirm the action.



In [8]:
df.drop_duplicates(inplace=True)
df.dropna(subset=['3P%'], inplace=True)
display(df.isnull().sum())

Name           0
GP             0
MIN            0
PTS            0
FGM            0
FGA            0
FG%            0
3P Made        0
3PA            0
3P%            0
FTM            0
FTA            0
FT%            0
OREB           0
DREB           0
REB            0
AST            0
STL            0
BLK            0
TOV            0
TARGET_5Yrs    0
dtype: int64

**Reasoning**:
Check for outliers in numerical columns using descriptive statistics and decide if any treatment is necessary based on extreme values.



In [9]:
display(df.describe())

GP          MIN          PTS          FGM          FGA  \
count  1318.000000  1318.000000  1318.000000  1318.000000  1318.000000   
mean     60.402124    17.654249     6.821017     2.635357     5.900910   
std      17.457767     8.340224     4.368868     1.687763     3.600133   
min      11.000000     3.100000     0.700000     0.300000     0.800000   
25%      47.000000    10.825000     3.700000     1.400000     3.300000   
50%      63.000000    16.100000     5.600000     2.100000     4.800000   
75%      77.000000    22.975000     8.800000     3.400000     7.500000   
max      82.000000    40.900000    28.200000    10.200000    19.800000   

               FG%      3P Made          3PA          3P%          FTM  \
count  1318.000000  1318.000000  1318.000000  1318.000000  1318.000000   
mean     44.138998     0.249621     0.784598    19.322989     1.302731   
std       6.116819     0.385397     1.065365    16.045404     0.989255   
min      23.800000     0.000000     0.000000     0.000000     0.000000   
25%      40.200000     0.000000     0.000000     0.000000     0.600000   
50%      44.000000     0.100000     0.300000    22.450000     1.000000   
75%      47.800000     0.400000     1.200000    32.575000     1.700000   
max      73.700000     2.300000     6.500000   100.000000     7.700000   

               FTA          FT%         OREB         DREB          REB  \
count  1318.000000  1318.000000  1318.000000  1318.000000  1318.000000   
mean      1.827162    70.412822     1.008194     2.023596     3.030956   
std       1.325931    10.494442     0.779410     1.364257     2.064212   
min       0.000000     0.000000     0.000000     0.200000     0.300000   
25%       0.900000    64.800000     0.400000     1.000000     1.500000   
50%       1.500000    71.300000     0.800000     1.700000     2.500000   
75%       2.300000    77.600000     1.400000     2.600000     3.975000   
max      10.200000   100.000000     5.300000     9.600000    13.900000   

               AST          STL          BLK          TOV  TARGET_5Yrs  
count  1318.000000  1318.000000  1318.000000  1318.000000  1318.000000  
mean      1.558801     0.620637     0.366161     1.195827     0.621396  
std       1.478618     0.411766     0.429008     0.725690     0.485223  
min       0.000000     0.000000     0.000000     0.100000     0.000000  
25%       0.600000     0.300000     0.100000     0.700000     0.000000  
50%       1.100000     0.500000     0.200000     1.000000     1.000000  
75%       2.000000     0.800000     0.500000     1.500000     1.000000  
max      10.600000     2.500000     3.900000     4.400000     1.000000

## 資料整理與分析

### Subtask:
根據您的具體需求，進行資料分組、聚合、轉換或執行更深入的分析（例如相關性分析、分佈分析等）。


**Reasoning**:
Calculate and display the correlation matrix for numerical columns, calculate and display correlations with 'TARGET_5Yrs', calculate and display the distribution of 'TARGET_5Yrs', and calculate and display the mean of selected numerical columns grouped by 'TARGET_5Yrs'.



In [10]:
display(df.corr(numeric_only=True))

display(df.corr(numeric_only=True)['TARGET_5Yrs'].sort_values(ascending=False))

display(df['TARGET_5Yrs'].value_counts())

selected_cols = ['PTS', 'AST', 'REB']
display(df.groupby('TARGET_5Yrs')[selected_cols].mean())

GP       MIN       PTS       FGM       FGA       FG%  \
GP           1.000000  0.593356  0.541334  0.545647  0.521358  0.293458   
MIN          0.593356  1.000000  0.911624  0.902738  0.910329  0.207097   
PTS          0.541334  0.911624  1.000000  0.990867  0.980041  0.259893   
FGM          0.545647  0.902738  0.990867  1.000000  0.980315  0.295872   
FGA          0.521358  0.910329  0.980041  0.980315  1.000000  0.135322   
FG%          0.293458  0.207097  0.259893  0.295872  0.135322  1.000000   
3P Made      0.111331  0.391862  0.348517  0.290943  0.391395 -0.289375   
3PA          0.103249  0.405573  0.358777  0.301147  0.414618 -0.345068   
3P%          0.041476  0.168600  0.154641  0.122794  0.199887 -0.328295   
FTM          0.483782  0.791200  0.896424  0.848348  0.827497  0.249928   
FTA          0.481036  0.779971  0.881036  0.840949  0.806829  0.303066   
FT%          0.193759  0.236707  0.254899  0.219750  0.265184 -0.153853   
OREB         0.401362  0.571772  0.573848  0.594985  0.503012  0.511662   
DREB         0.467714  0.746550  0.695453  0.704316  0.642465  0.409057   
REB          0.461048  0.709949  0.677414  0.691271  0.615463  0.464560   
AST          0.374063  0.630508  0.553357  0.534055  0.591089 -0.104841   
STL          0.454288  0.756955  0.674935  0.662483  0.690166  0.059482   
BLK          0.278236  0.398896  0.387611  0.398186  0.323084  0.390073   
TOV          0.520439  0.826235  0.850256  0.834340  0.846062  0.124861   
TARGET_5Yrs  0.403781  0.318435  0.315697  0.317226  0.291205  0.237833   

              3P Made       3PA       3P%       FTM       FTA       FT%  \
GP           0.111331  0.103249  0.041476  0.483782  0.481036  0.193759   
MIN          0.391862  0.405573  0.168600  0.791200  0.779971  0.236707   
PTS          0.348517  0.358777  0.154641  0.896424  0.881036  0.254899   
FGM          0.290943  0.301147  0.122794  0.848348  0.840949  0.219750   
FGA          0.391395  0.414618  0.199887  0.827497  0.806829  0.265184   
FG%         -0.289375 -0.345068 -0.328295  0.249928  0.303066 -0.153853   
3P Made      1.000000  0.982640  0.588538  0.159747  0.097019  0.314893   
3PA          0.982640  1.000000  0.580918  0.175176  0.110445  0.323614   
3P%          0.588538  0.580918  1.000000  0.035144 -0.027449  0.328961   
FTM          0.159747  0.175176  0.035144  1.000000  0.980535  0.253283   
FTA          0.097019  0.110445 -0.027449  0.980535  1.000000  0.110592   
FT%          0.314893  0.323614  0.328961  0.253283  0.110592  1.000000   
OREB        -0.217329 -0.230110 -0.287360  0.584242  0.653664 -0.148935   
DREB         0.020387  0.015803 -0.120756  0.656246  0.702989 -0.023008   
REB         -0.069297 -0.077197 -0.189001  0.655569  0.712897 -0.072043   
AST          0.375183  0.409049  0.263488  0.476469  0.429291  0.292118   
STL          0.306702  0.338800  0.196523  0.599654  0.579909  0.203417   
BLK         -0.154842 -0.168020 -0.240578  0.408944  0.470118 -0.160810   
TOV          0.257924  0.283445  0.110138  0.805715  0.799897  0.195364   
TARGET_5Yrs  0.036384  0.017500 -0.003065  0.297329  0.298180  0.096356   

                 OREB      DREB       REB       AST       STL       BLK  \
GP           0.401362  0.467714  0.461048  0.374063  0.454288  0.278236   
MIN          0.571772  0.746550  0.709949  0.630508  0.756955  0.398896   
PTS          0.573848  0.695453  0.677414  0.553357  0.674935  0.387611   
FGM          0.594985  0.704316  0.691271  0.534055  0.662483  0.398186   
FGA          0.503012  0.642465  0.615463  0.591089  0.690166  0.323084   
FG%          0.511662  0.409057  0.464560 -0.104841  0.059482  0.390073   
3P Made     -0.217329  0.020387 -0.069297  0.375183  0.306702 -0.154842   
3PA         -0.230110  0.015803 -0.077197  0.409049  0.338800 -0.168020   
3P%         -0.287360 -0.120756 -0.189001  0.263488  0.196523 -0.240578   
FTM          0.584242  0.656246  0.655569  0.476469  0.599654  0.408944   
FTA          0.653664  0.702989  0.712897 

TARGET_5Yrs    1.000000
GP             0.403781
MIN            0.318435
FGM            0.317226
PTS            0.315697
REB            0.301905
FTA            0.298180
FTM            0.297329
OREB           0.295316
FGA            0.291205
DREB           0.287532
TOV            0.273680
FG%            0.237833
STL            0.229533
BLK            0.212342
AST            0.177313
FT%            0.096356
3P Made        0.036384
3PA            0.017500
3P%           -0.003065
Name: TARGET_5Yrs, dtype: float64

TARGET_5Yrs
1    819
0    499
Name: count, dtype: int64

PTS       AST       REB
TARGET_5Yrs                              
0            5.054709  1.223046  2.232866
1            7.897192  1.763370  3.517216

## 結果呈現

### Subtask:
以清晰易懂的方式呈現分析結果，例如使用表格、圖表或文字說明。


## Summary:

### Data Analysis Key Findings

*   The dataset contains 1340 entries and 21 columns, including player statistics and a target variable `TARGET_5Yrs`.
*   The `TARGET_5Yrs` column indicates whether a player stayed in the league for 5 years or more (1) or not (0).
*   The `3P%` column was identified as having 11 missing values.
*   Duplicate rows were successfully removed from the dataset.
*   Rows with missing values in the `3P%` column were removed, resulting in a clean dataset for analysis.
*   Based on descriptive statistics, no extreme outliers were identified in the numerical columns.
*   `GP` (Games Played) showed the strongest positive correlation with `TARGET_5Yrs` (approximately 0.40).
*   Other statistics like MIN, FGM, PTS, REB, FTA, FTM, OREB, FGA, DREB, and TOV also had positive correlations with `TARGET_5Yrs`.
*   Players who stayed for 5 years or more (`TARGET_5Yrs` = 1) had higher average PTS, AST, and REB compared to those who did not (`TARGET_5Yrs` = 0).

### Insights or Next Steps

*   Consider building a predictive model using features with strong correlations to `TARGET_5Yrs` (like GP, MIN, PTS, AST, REB) to predict player longevity.
*   Explore the distribution of key features for each `TARGET_5Yrs` group using visualizations (e.g., histograms, box plots) to gain further insights into performance differences.
